url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/org/jgraph/graph/GraphTransferHandler.java

```diff
@@ -0,0 +1,382 @@
+import org.jgraph.JGraph;
+
+/**
+ * @author Gaudenz Alder
+ * 
+ * Default datatransfer handler.
+ */
+public class GraphTransferHandler extends TransferHandler {
+
+	/**
+	 * Controls if all inserts should be handled as external drops even if all
+	 * cells are already in the graph model. This is useful if the enclosing
+	 * component does not allow moving.
+	 */
+	protected boolean alwaysReceiveAsCopyAction = false;
+
+	/* Pointer to the last inserted array of cells. */
+	protected Object out, in;
+
+	/* How many times the last transferable was inserted. */
+	protected int inCount = 0;
+
+	public boolean canImport(JComponent comp, DataFlavor[] flavors) {
+		for (int i = 0; i < flavors.length; i++)
+			if (flavors[i] == GraphTransferable.dataFlavor)
+				return true;
+		return false;
+	}
+
+	/* Public entry point to create a Transferable. */
+	public Transferable createTransferableForGraph(JGraph graph) {
+		return createTransferable(graph);
+	}
+
+	protected Transferable createTransferable(JComponent c) {
+		if (c instanceof JGraph) {
+			JGraph graph = (JGraph) c;
+			if (!graph.isSelectionEmpty()) {
+				return createTransferable(graph, graph.getSelectionCells());
+			}
+		}
+		return null;
+	}
+
+	protected Transferable createTransferable(JGraph graph, Object[] cells) {
+		Object[] flat = graph.getDescendants(graph.order(cells));
+		ParentMap pm = ParentMap.create(graph.getModel(), flat, false, true);
+		ConnectionSet cs = ConnectionSet.create(graph.getModel(), flat, false);
+		Map viewAttributes = GraphConstants.createAttributes(flat, graph
+				.getGraphLayoutCache());
+		Rectangle2D bounds = graph.getCellBounds(graph.getSelectionCells());
+		bounds = new AttributeMap.SerializableRectangle2D(bounds.getX(), bounds
+				.getY(), bounds.getWidth(), bounds.getHeight());
+		out = flat;
+		return create(graph, flat, viewAttributes, bounds, cs, pm);
+	}
+
+	protected GraphTransferable create(JGraph graph, Object[] cells,
+			Map viewAttributes, Rectangle2D bounds, ConnectionSet cs,
+			ParentMap pm) {
+		return new GraphTransferable(cells, viewAttributes, bounds, cs, pm);
+	}
+
+	protected void exportDone(JComponent comp, Transferable data, int action) {
+		if (comp instanceof JGraph && data instanceof GraphTransferable) {
+			JGraph graph = (JGraph) comp;
+			if (action == TransferHandler.MOVE) {
+				Object[] cells = ((GraphTransferable) data).getCells();
+				graph.getGraphLayoutCache().remove(cells);
+			}
+			graph.getUI().updateHandle();
+			graph.getUI().setInsertionLocation(null);
+		}
+	}
+
+	public int getSourceActions(JComponent c) {
+		return COPY_OR_MOVE;
+	}
+
+	// NOTE: 1. We abuse return value to signal removal to the sender.
+	// 2. We always clone cells when transferred between two models
+	// This is because they contain parts of the model's data.
+	// 3. Transfer is passed to importDataImpl for unsupported
+	// dataflavors (becaue method may return false, see 1.)
+	public boolean importData(JComponent comp, Transferable t) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 37. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.